(feat) added unsubscribe functionality + new documentation (README)

simku22 opened the pull request:
<!--
Thanks for contributing to a BobaTalks project! Please take some time to fill out the following fields to provide context to reviewers and other collaborators:
-->

## Description
#### What is the purpose of your change?
- To allow the user to unsubscribe from selected tags in the "resumes" channel.

#### What areas of the application does this change impact? (changes API, folder structure, etc)
- Changes slash command (adds new), Menu.py, and the files in server (routes.py, database.py). It creates a new route as well as new commands to fetch subscribed tags and remove subscribers.

#### What features/fixes are deferred/not handled in this PR?
- Ability to select multiple unsubscribe options at once and handle errors where the users attempts to unsubscribe from an already unsubscribed id. Otherwise this works efficiently for single tag subs/unsubs.

## Checklist
- [X] Has a work item linked to the PR

simku22 committed:
Improved README instructions

- added instructors for creating a new discord bot and server
- improved ordering of text elements for readability
- added guide on setting up docker and mongodb
- added guide for running test environment for the first time

Co-Authored-By: Anthony Camarillo <58638902+a-camarillo@users.noreply.github.com>

simku22 committed:
v1 of unsubscribe feature

simku22 committed:
working fetch subscribed tags implemented

able to find out what users can "unsubscribe" from by returning a list of all the tags they are currently subscribed to

simku22 committed:
working unsubscribe feature (needs testing)

- added a main slash command in bot.py
- added a boolean to Menu.py, so that we can specify whether a subscribe/unsubscribe operation is happening
- small linting stuff from the black pre-commit thing

simku22 committed:
Merge pull request #1 from simku22/simku22-testing-branch

working sub/unsub feature + readme updates + misc changes

jwu910 commented on .gitignore:
```diff
@@ -3,3 +3,20 @@ venv
 __pycache__
 
 .env
+
+.env.test
```
shouldn't `.env.test` be caught by `.env`? or can we just add `*` to the end of `.env`? no need for multiple ones.

simku22 replied:
I think this was a file error from my end, this and the node packages were added accidentally.

simku22 replied:
Will remove this.

jwu910 commented on .gitignore:
```diff
@@ -3,3 +3,20 @@ venv
+.env.test
+
+package.json
+package-lock.json
```
why do we have npm stuff here?

jwu910 commented on .gitignore:
```diff
@@ -3,3 +3,20 @@ venv
+node_modules/dotenv/config.js
+node_modules/dotenv/lib/cli-options.js
+node_modules/dotenv/lib/env-options.js
+vars.env
```
I think we need to delete some stuff. we're not running a node project here, we shouldn't need to ignore these files.

jwu910 commented on README.md:
```diff
@@ -2,34 +2,142 @@
+
+Invite your bot to your server with the following set of permissions:
+
+<div style="padding-bottom: 20px;">
+    <li> Manage Server
+    <li> Manage Channels
+    <li> Send Messages
+    <li> Create Public Threads
+    <li> Manage Messages
+    <li> Embed Links
+    <li> Use Slash Commands
+</div>
```
markdown syntax `-` prefixes might be enough for bullet points

simku22 replied:
![image](https://github.com/BobaTalks/bobabot/assets/116035879/1b41344e-1de0-4156-9f20-ad5b87e4f9d1)

![Capture](https://github.com/BobaTalks/bobabot/assets/116035879/991a2679-ecd0-4239-8966-8e0831c72d40)

I think this doesn't format entirely correctly in markdown since its in a div. Will keep as  < li> for now, but can swap for list in markdown as well.

a-camarillo replied:
yah not entirely sure what the purpose of a div is here

jwu910 commented on README.md:
````diff
@@ -2,34 +2,142 @@
+Pull MongoDB
+
+```shell
+docker pull mongo:latest
````
do we want to establish a version of mongo for consistency?

simku22 replied:
will talk to anthony abt this

a-camarillo replied:
uh idk, I suppose we could but I'm not sure if it's really necessary

jwu910 commented on bot/bot.py:
```diff
@@ -68,6 +68,20 @@ def create_message_string(iter_tags):
     return ", ".join(tags)
 
 
+@bot.command()
+async def boba(ctx):
+    """
+    Creates a boba-ascii art and outputs to respective channel
+
+    Parameters
+    ----------
+    ctx : Discord Context
+        Includes information on who executed the command (namely channel)
+    """
+    ascii_art = "\n⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢿⡄⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢸⣇⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣿⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⣀⣠⣤⣤⣤⣤⣴⣶⣶⣦⣤⣤⣤⣤⣄⣀⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠉⣉⠉⠉⠹⣿⠀⠀⠀⠀⠀⠈⠉⠉⣉⠉⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⢿⡇⠀⠀⠉⠀⠀⠀⠀⠀⠀⠀⢸⡿⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⢸⣿⣿⣿⡆⠘⣿⣿⣿⣿⣿⣿⣿⡇⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⢸⣿⣿⣿⣷⠀⢻⣿⣿⣿⣿⣿⣿⡇⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⠈⣿⣿⣿⣿⡄⠸⣿⣿⣿⣿⣿⣿⠁⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⠀⣿⣿⣿⣿⣧⠀⢿⣿⣿⡟⢻⣿⠀⠀⠀⠀⠀⠀⠀⠀⠀\n⠀⠀⠀⠀⠀⠀⠀⠀⠀⣿⡿⣿⠉⢻⡀⢸⡏⢉⡗⢺⡿⠀⠀⠀⠀⠀⠀⠀ ...
+    await ctx.send(ascii_art)
+
+
```
i don't think this is in scope, can we pull this out and open a ticket if we need this?

simku22 replied:
yes, will remove.

jwu910 commented on bot/client_requests.py:
```diff
@@ -49,3 +50,41 @@ def add_subscriber(user_id, tag_id):
+    requests.delete(f"{server_url}/tags/{tag_id}/subscribers/{user_id}")
+
+
+def fetch_subscriptions(user_id):
```
can we call this something like `fetch_ubscriptions_by_user_id` ?

jwu910 commented on bot/client_requests.py:
```diff
@@ -49,3 +50,41 @@ def add_subscriber(user_id, tag_id):
+
+    Parameters
+    ----------
+    user_id : int
+        An integer representing the corresponding user id
```
TIL python int has no max size

jwu910 commented on bot/menu.py:
```diff
@@ -18,13 +18,14 @@ class Menu(Select):
     """
 
-    def __init__(self, tags):
+    def __init__(self, tags, sub_toggle):
```
can we minimize abbreviations?

simku22 replied:
renamed this to `is_adding_subscriber`

jwu910 commented on bot/menu.py:
```diff
@@ -48,8 +49,12 @@ async def callback(self, interaction: Interaction):
         """
-        for value in self.values:
-            add_subscriber(interaction.user.id, value)
+        if self.sub_toggle:
+            for value in self.values:
+                add_subscriber(interaction.user.id, value)
+        else:
+            for value in self.values:
+                remove_subscriber(interaction.user.id, value)
```
Not familiar enough to question this directly, and it might be just the frontend background thinking about this, but putting a new list into the DB of subscriptions vs finding and removing individual subscriptions sounds cleaner

but up to you and @a-camarillo on this pattern i think

jwu910 commented on bot/menu.py:
```diff
@@ -78,7 +86,7 @@ def add_menu(self, tags):
             The tags owned by the Forum channel
         """
-        menu = Menu(tags)
+        menu = Menu(tags, self.sub_toggle)
```
Does the menu not allow us to add/remove within the menu like a series of available tags and a switch to turn them on/off? 

I dont recall what the menu looks like right now, got any screenshots as well?

simku22 replied:
![image](https://github.com/BobaTalks/bobabot/assets/116035879/910406d3-310e-4048-8edf-857eccdd7b25)
The way this currently works is like an on/off switch for the callback in the menu. It allows us to use the same menu object and pass in a simple set of tags but just to specify whether we want to add or remove subscribers.

jwu910 commented on server/database.py:
```diff
@@ -79,6 +79,16 @@ def create_subscriber(db, tag_id, subscriber_id):
 
 
+# (IP) remove_subscriber removes a subscriber from a selected tag id in a database
+def remove_subscriber_from_tag(db, tag_id, subscriber_id):
+    query_results = db.tags.update_one(
+        {"snowflake_id": tag_id}, {"$pull": {"subscribers": subscriber_id}}
+    )
+    if query_results.modified_count == 0:
+        return ({"Error": "Subscriber Does Not Exist"}, 404)
+    return ({"Success": "Delete Successful"}, 200)
+
+
```
where should we expect to wrap our error handling?

simku22 replied:
I might be misinterpreting, but currently this follows the format for the other operations in database.py it provides an error code if there is no change to the subscriber list -> which means we either didn't successfully fetch the list or  some other connection error.

a-camarillo replied:
Correct. The way we currently have things configured, functions in database.py act as queries to the db. The query functions return a status code and data if data is to be expected i.e. a Read or GET query. What to do with that data is then the responsibility of the flask server.

simku22 committed:
addressed change_requests from @jwu910

simku22 commented:
https://github.com/BobaTalks/bobabot/pull/23/commits/c464bdf1bdf724cf7a5b1cbb2a0ac45c90776043

simku22 committed:
removing div from readme

small bug fix for docs

simku22 commented:
d6280e488da036f940af0ba9cd36ece83c1d1327

jwu910 commented on README.md:
````diff
@@ -2,34 +2,140 @@
+Within your repository, create a file named ".env" with the following
+
+```shell
+# Port should match your container
+MONGO_URI='mongodb://localhost:27017/'
+# Database name is your container name
+DATABASE_NAME='bobabot'
+# To create a token, register on discord for developers. Create a new application, and see under the bot tab. Copy paste your token here.
+DISCORD_TOKEN='abCd214sampleToken'
+# Set to the name of your testing discord
+DISCORD_SERVER_NAME='sample_server_name'
+# Set to the name of your testing community "forum" channel
+DISCORD_CHANNEL_NAME='resumes'
+SERVER_URL='http://localhost:5000'
````
Can we commit this to a .env.sample or sample.env? Then we can leave this out of here, env files are pretty common now and we can reduce the instructions to copying the sample file to `.env` and add in any secrets there

simku22 replied:
I can do that, but I don't think it would change much. I think generally this wouldn't change the onboarding or testing flow at all, the person still has to copy it from one location to another

jwu910 commented on server/database.py:
```diff
@@ -79,6 +79,16 @@ def create_subscriber(db, tag_id, subscriber_id):
     return ({"Success": "Resource Created"}, 201)
 
 
+# (IP) remove_subscriber removes a subscriber from a selected tag id in a database
```
What's (IP)? Can we use the same docstring format we're using with other methods?

simku22 replied:
IP just means in progress, I test in discord this until I run it alongside anthony's notification system. In progress is inherent here, so I'll remove it.

simku22 committed:
readme + database.py fixes

- removed the .env contents from the readme and put it into sample.env
- made comments on new methods consistent in database.py

simku22 commented:
[882564c](https://github.com/BobaTalks/bobabot/pull/23/commits/882564c23e97b1f9f1509349a412497386d27905)

This addresses both of these CR's.

jwu910 commented:
@simku22 can you resolve conflicts and we'll take another look?

simku22 committed:
Merge branch 'main' into main

simku22 committed:
(chore) fixed white spacing

- ran linter
